### PR TITLE
tests: simplify libvmi and use it to start diskless VM

### PR DESCRIPTION
**Pull request by dankenigsberg**
This PR purges two public constants from `tests.libvmi`, as they are not required and dropping them makes the code simpler. In particular, it makes it simpler to start a diskless VM via `libvmi` and simplifies the code introduced in PR #7649.

/cc @EdDev as this modifies `libvmi` APIs
/cc @enp0s3 as this modifies `infra_test`'s rate limiter test

```release-note
NONE
```

### CODE REVIEW

**Comment by codingben** [comment]
JFYI, I see similar failures of `pull-kubevirt-fossa` in other PRs. Looks like we need to update it:
```
fossas/fossa-cli crit You are installing FOSSA CLI v1, which is no longer in active
fossas/fossa-cli crit development. FOSSA will not address new defects found in CLI v1.
```

File: tests/libvmi/vmi.go
Comment (EdDev) [inline tests/libvmi/vmi.go:132]: > Users of libvmi do not to know about libvmi.DefaultTestGracePeriod. If
> they care about the grace period of their VMs, they can set it with
> WithTerminationGracePeriod(). Nobody does that, so exporting
> WithTerminationGracePeriod is probably premature.

I do not know what is the meaning of assigning a zero grace period vs keeping it a `nil`. I think the original idea with the default was to copy the existing settings from the code, but now that you want to move it, maybe it is worth asking why we need to set it to zero in the first place.

Also, if you remove all usages of `WithTerminationGracePeriod()`, I see no other option but to remove that builder function as well, otherwise it is left as dead code.
I do find it a bit inaccurate to set this default here so deep. All other settings of the `baseVMI` are a bare minimum... is this the same?

File: tests/infra_test.go
Comment (EdDev) [inline tests/infra_test.go:166]: Just a nit about formatting:
```
vmi := libvmi.New(
    libvmi.RandName(),
    libvmi.WithResourceMemory("1Mi"),
    libvmi.WithNodeSelectorFor(&targetNode),
)
```

**Review by EdDev — COMMENTED** [review]
Thanks. Overall I am good with this change, there is only one issue I prefer resolved before ack-ing it (see inline).

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go:132]: > > Users of libvmi do not to know about libvmi.DefaultTestGracePeriod. If
> > they care about the grace period of their VMs, they can set it with
> > WithTerminationGracePeriod(). Nobody does that, so exporting
> > WithTerminationGracePeriod is probably premature.
> 
> I do not know what is the meaning of assigning a zero grace period vs keeping it a `nil`. I think the original idea with the default was to copy the existing settings from the code, but now that you want to move it, maybe it is worth asking why we need to set it to zero in the first place.

The [server-side default is 30s](https://github.com/kubevirt/kubevirt/blob/dd649a1ca34/staging/src/kubevirt.io/api/core/v1/types.go#L39). I assumed that we set it to 0s in the tests because we want the tests to run faster, and we do not care about anything in the test vm to shut down cleanly. Commit 8c72656a52 changed it from 1s to 0s and I think we should keep it.

> 
> Also, if you remove all usages of `WithTerminationGracePeriod()`, I see no other option but to remove that builder function as well, otherwise it is left as dead code.

I agree that it is currently left dead, but we make a strong push to move all code to `libvmi`, and we do have a test for the termination grace period in `tests/vmlifecycle_test.go`. It should not be long until we reintroduce a function like this, hence my reluctance to drop it now. But of course, I'll drop the function if you insist.

> I do find it a bit inaccurate to set this default here so deep. All other settings of the `baseVMI` are a bare minimum... is this the same?

Do you think that it belongs better in `New()`? I don't really mind. I pushed it down because 99% of the tests don't need to know about it.

Comment (EdDev) [reply tests/libvmi/vmi.go:132]: Appending to the `opts` in `New` may make it a bit nicer and cause the function to be actually used.
My main concern here is that we will start seeing other things added this way to that base function, which was not the intention. It will make it messy IMO, as we start adding defaults all over the stack, like the previous version of the VMI building.

Said that, I do not think it is worth overthinking this. It is good enough.

**Review by EdDev — APPROVED** [review]

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go:132]: > Appending to the `opts` in `New` may make it a bit nicer and cause the function to be actually used. 

Ok, but it needs to be pre-pended, to make future override possible.

**Review by EdDev — APPROVED** [review]
Thank you!

File: tests/libvmi/vmi.go
Comment (xpivarc) [inline tests/libvmi/vmi.go:41]: Why do we force it here? Are we sure it is safe to introduce it this late?

File: tests/network/vmi_slirp_interface.go
Comment (xpivarc) [inline tests/network/vmi_slirp_interface.go:197]: Maybe we could transform this to use libvmi and don't use the util functions...

**Review by xpivarc — COMMENTED** [review]
/hold

I have just a question about the teminationGrecePeriod.

File: tests/network/vmi_slirp_interface.go
Comment (dankenigsberg) [reply tests/network/vmi_slirp_interface.go:197]: I believe we could and we should - but not in this PR, please.
This commit is related to the PR only in that it drops a usage of an "evil" public constant (`util.NamespaceTestDefault`) but if you insist, I would kick it out of the PR.

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go:41]: I attempted to explain why we enforce this in the commit message:

> Users of libvmi should not to know about libvmi.DefaultTestGracePeriod. If
> they care about the grace period of their VMs, they can set it with
> WithTerminationGracePeriod(). We use WithTerminationGracePeriod(0) to
> make shutdown faster in the test. Once we migrate the termination grace
> period test to libvmi we would start using this function externally so
> we keep it exposed by the package.

I can try to rephrase: in all current usages of `libvmi`, we use `WithTerminationGracePeriod(0)` because we want a quick shutdown. A future test would want to override this value to test the `terminationGracePeriod` field.

What do you mean by "this late" ?

File: tests/network/vmi_slirp_interface.go
Comment (xpivarc) [reply tests/network/vmi_slirp_interface.go:197]: I should start indicating when I require a change. This is not one of those... I apologize

File: tests/libvmi/vmi.go
Comment (xpivarc) [reply tests/libvmi/vmi.go:41]: I understand and agree but we should not move it to `New` as it is now a bigger change than it has to be. I personally believe that now one(reviewer/approver) should review all the `New` usage to ensure the usage of `New` stays as intended.

Also, I would expect we don't add any option to `New` as imho it should be used for 1% of the use cases. The user should specify the intended spec and we should not force any smart optimization from our side.

Comment (xpivarc) [reply tests/libvmi/vmi.go:41]: By "late" I meant the fact we introduced this API and it is used. The author of tests or editor of tests could assume this method doesn't add any option and therefore it is safe to not be explicit about the `TerminationGracePeriod`

Comment (EdDev) [reply tests/libvmi/vmi.go:41]: @xpivarc , this was a default used in all factories and moved deeper to assure our tests will not use (by mistake) the system default.
I think this fits 99% of the existing and future usages, it will be an exception to see a different value set (including `nil`).

In reality, I would expect no one to care about this value, making this change accepteable.
I do understand that it is better to be explicit, and I would generally agree, but in this case the reasoning that @dankenigsberg gave regarding the exposure of default constants convinced me. Reducing the API options has a value, and this specific option does make sense to keep as an overall default.

Comment (xpivarc) [reply tests/libvmi/vmi.go:41]: @EdDev I understand this does make sense for factories but this is not a factory, right? I was surprised that there is no usage of `New` outside of factories. I would still argue that `New` should be "clean" and without any options. WDYT?

Comment (dankenigsberg) [reply tests/libvmi/vmi.go:41]: Originally I placed this assignment even deeper, in `baseVmi`, but @EdDev asked to take it a bit higher - apparently for similar reasoning. I believe that it is OK that KubeVirt has one default (30s) and the test-purposed library has another default (0s). They have a different audience and a different typical use case. I think that it would be wrong to force each and every present and future caller of `libvmi.New` to be aware of this difference, and I don't see any other way to avoid this if not by hiding it in an exported function.

Comment (dankenigsberg) [reply tests/libvmi/vmi.go:41]: > By "late" I meant the fact we introduced this API and it is used. The author of tests or editor of tests could assume this method doesn't add any option and therefore it is safe to not be explicit about the `TerminationGracePerio

The grand majority of tests use `NewAlpine` etc, where `TerminationGracePeriod` is *already* implicitly set to zero. My PR only levels the ground, so that the very few users of vanilla `New` has the same.

Comment (xpivarc) [reply tests/libvmi/vmi.go:41]: > 

I am not sure why it is more ok to have it `New` than `baseVMI`. I am going to unhold as there is no usage of `New` outside of factories but I believe it is wrong.

Comment (dankenigsberg) [reply tests/libvmi/vmi.go:41]: I'm trying to understand what exactly you find wrong. Tests having a default that is different from what the production code has?

Comment (xpivarc) [reply tests/libvmi/vmi.go:41]: Having a default in the generic constructor.

Comment (dankenigsberg) [reply tests/libvmi/vmi.go:41]: @xpivarc so how about keeping the generic `New` pristine, and exposing a fresh factory `NewDiskless` to be used by tests? I'm not sure myself why it would be better, but name-wise it is more in line with `NewAlpine`, `NewFedora` etc. I suspect that it many other tests using `NewDiskless` would speed up boot and shutdown without any cost to the test case.

Comment (EdDev) [reply tests/libvmi/vmi.go:41]: First I find it (very) good to see such an argument (compared to ignoring quality in many other reviews at other PR/s).

To this topic, I think there is no one best solution, there will be always pros and cons.
I think the current code is good enough, even though it got added a special default. But I am also fine with Dan's proposal, although it has the cons of not being used much (therefore less a fit under `libvmi` IMO).

There is so much more to do to cleanup the code-base, lets pick up something bigger and tackle it. If one of you will prefer something else/better here, please post a PR and we can talk about it there.

**Pull request merged**
